syntax = "proto3";


message EchoRequest {
  bytes message = 1;
}

message EchoResponse {
  bytes message = 1;  // bytes -> base64 encoded
  bool status = 2;
}

message MoveRequest {
  uint32 step = 1;
}

message Status {
  bool status = 1;
}

message EmptyRequest {
}

message TurnRequest {
  uint32 angle = 1;
}

message RobotMode {
  enum Mode {
    EXPLORATION = 0;
    FASTEST_PATH = 1;
  }

  Mode mode = 1;
}

message MetricResponse {
  map<uint32, float> values = 1;
  bool status = 2;
}

message Position {
  enum Direction {
    NORTH = 0;
    EAST = 1;
    SOUTH = 2;
    WEST = 3;
  }

  int32 x = 1;
  int32 y = 2;
  Direction dir = 3;
}

message MapDescription {
  string p1 = 1;  // Explored map description
  string p2 = 2;  // Obstacle map description
}

message RobotInfo {
  Position pos = 1;
  MapDescription map = 2;
  repeated ImagePosition images = 3;
  RobotStatus robot_status = 4;
}

enum RobotStatus {
  STOP = 0;
  FORWARD = 1;
  TURN_LEFT = 2;
  TURN_RIGHT = 3;
}

message ImagePosition {
  enum ImageID {
    UNKNOWN = 0;  // skip 0
    WHITE_UP = 1;
    BLUE_DOWN = 2;
    YELLOW_RIGHT = 3;
    RED_LEFT = 4;
    GREEN_CIRCLE = 5;
    BLUE_SIX = 6;
    GREEN_SEVEN = 7;
    RED_EIGHT = 8;
    WHITE_NINE = 9;
    YELLOW_ZERO = 10;
    RED_V = 11;
    GREEN_W = 12;
    WHITE_X = 13;
    BLUE_Y = 14;
    YELLOW_Z = 15;
  }
  uint32 x = 1;
  uint32 y = 2;
  ImageID id = 3;
}

message ImageResponse {
  bytes raw_image = 1;
}

message CalibrationRequest {
  // mode 0 = front
  // mode 1 = wall
  int32 mode = 1;
}
